### PR TITLE
#410 - Retrieve navigation page names and slugs programmatically

### PR DESCRIPTION
**Pull request by blms**
## What this PR does

- Per #410:
  - Creates a `ContainerPage` class of which `ContentPage` instances may be subpages
  - Updates the bootstrap script to create an `ContainerPage` instance and make all `ContentPage` instances subpages of that instance, except for "Contact Us" and the test fixture page
  - Updates the nav template to use the new logic to display children of the Home page in the main nav, and children of Container pages in their sub-navs

## Deploy Notes

See `DEPLOYNOTES.md`. In addition:

- Content editors should be made aware that to edit pages in sub-menus, in the Wagtail admin, they need to click the green right arrow button next to a Container page. Editing the Container page itself will only allow them to change its title or slug.
- Content editors should also be made aware that new content pages must be of type "Content page", regardless of whether they are to appear in the main navigation or a sub-menu. If they want the new page to appear in a sub-menu, they should be made child pages of an existing "Container page" (like the others in that sub-menu).

### CODE REVIEW

File: geniza/pages/models.py [hunk 1]
```diff
@@ -39,6 +40,19 @@ class Meta:
...
+    """An empty :class:`Page` type that has :class:`ContentPage` instances
+    as its subpages."""
+
+    # can only be child of HomePage
+    parent_page_types = [HomePage]
+    subpage_types = ["pages.ContentPage"]
+
+    # should not ever actually render
+    def serve(self, _):
+        raise Http404
```
[the remark this answers is not in the document]
Comment (rlskoeser) [reply geniza/pages/models.py:53]: Maybe it should redirect to the parent page.

File: geniza/templates/nav.html [hunk 2]
```diff
@@ -66,33 +68,21 @@
...
-                    </li>
-                    <li class="menu-item" role="none">
-                        {# Translators: label for link to 'FAQ' page in main navigation #}
-                        <a href="/faq/" role="menuitem">{% translate "FAQ" %}</a>
-                    </li>
+                    {% for child_page in about_page.localized.get_children.live.in_menu %}
+                        <li class="menu-item" role="none">
+                            {# Translators: label for link to 'Credits' page in main navigation #}
+                            <a href="/{{ about_page.localized.slug }}/{{ child_page.localized.slug }}/" role="menuitem">
+                                {{ child_page.localized.title }}
```
Comment (rlskoeser) [inline geniza/templates/nav.html:82]: We shouldn't be constructing the url like this.  Can't we use `page_url` instead?

Also, I think the translators comment is out of date now.

```suggestion
                            <a href="{% pageurl child_page.localized %}" role="menuitem">
                                {{ child_page.localized.title }}
```

File: geniza/templates/nav.html [hunk 3]
```diff
@@ -42,18 +42,20 @@
...
                 <a id="open-about-menu"
                     href="#about-menu"
                     aria-haspopup="true"
                     aria-expanded="false"
                     role="button">
-                    {{ about_label }}
+                    {{ about_page.localized.title }}
                 </a>
                 {# Translators: accessibility label for 'about' menu in navigation #}
                 {% translate "about menu" as about_menu_label %}
```
Comment (rlskoeser) [inline geniza/templates/nav.html:61]: Would like to see this not be hard-coded to the about page set in context, but just iterate over pages and handle a container page with child pages accordingly.

**Review by rlskoeser — CHANGES_REQUESTED** [review]
I would like to see `AboutPage` renamed to something more general and descriptive of how it functions, and see the nav logic reworked so it is based on page type and not hard-coded for current about page logic. (The about page should have `in_menus` set to true and nav should follow page order based on wagtail site.)

Not sure what to do about re-running the bootstrap content script — it doesn't look like they've added much content in production yet (just some preliminary credits), but it still might be easier to make the adjustments manually since we already have the locale mirroring configured. (Otherwise I think we have to delete all the sites, run the scripts, re-configure locales...)  Please add instructions to the deploy notes for the next version about what needs to be done to adjust for this.

File: geniza/pages/models.py [hunk 4]
```diff
@@ -39,6 +40,19 @@ class Meta:
         verbose_name = "homepage"
 
 
+class AboutPage(Page):
```
[the remark this answers is not in the document]
Comment (blms) [reply geniza/pages/models.py]: Done in a603fa0 👍

File: geniza/pages/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (blms) [reply geniza/pages/models.py]: Done in a603fa0 👍

File: geniza/context_processors.py [hunk 5]
```diff
@@ -1,12 +1,18 @@
...
 def template_globals(request):
     """Template context processor: add global includes (e.g.
     from django settings or site search form) for use on any page."""
 
+    site = Site.find_for_request(request)
     context_extras = {
         "SHOW_TEST_WARNING": getattr(settings, "SHOW_TEST_WARNING", False),
         "FONT_URL_PREFIX": getattr(settings, "FONT_URL_PREFIX", ""),
+        "site": site,
+        "about_page": site.root_page.get_children().exact_type(AboutPage).first(),
```
[the remark this answers is not in the document]
Comment (blms) [reply geniza/context_processors.py]: Removed in 1269ac1 👍

File: geniza/templates/nav.html (hunk 3 above)
Comment (blms) [reply geniza/templates/nav.html]: Done in 1269ac1 👍

**Comment by blms** [comment]
@rlskoeser Last thing, I do still wonder if it's a bit confusing when you create a child page of the homepage, where it might appear as though to create a page that appears in the submenu, one should click "Sub-menu page" here:

Screen Shot 2022-01-05 at 12 35 12 PM

I went with "sub menu page" to explain its display purpose, but maybe "container page" _would_ be a better name to avoid the confusion. What do you think?

File: geniza/templates/nav.html [hunk 6]
```diff
@@ -43,48 +43,49 @@
...
                     </li>
-                    <li class="menu-label" role="none">
-                        <span>{{ about_page.localized.title }}</span>
-                    </li>
-                    <li class="menu-button" role="none">
-                        <a href="#" role="button">
-                            <span class="sr-only">{{ close_button }}</span>
+                {% else %}
+                    <li class="menu-item" role="none">
+                        <a href="/{{ top_level_page.localized.slug }}/" role="menuitem">
```
Comment (rlskoeser) [inline geniza/templates/nav.html:83]: Let's use `pageurl` here too
```suggestion
                        <a href="{% pageurl child_page.localized %}" role="menuitem">
```

**Review by rlskoeser — APPROVED** [review]
Revisions & deploy notes look good to me. Just one change needed where we should use `pageurl`

File: geniza/templates/nav.html (hunk 6 above)
Comment (blms) [reply geniza/templates/nav.html]: Done in d6c9c58 👍

**Comment by blms** [comment]
Good catch! Also regenerated translations in 91428f6. Will merge when checks complete.

**Pull request merged**